jfr-mappers/src/main/java/com/newrelic/jfr/profiler/ProfileSummarizer.java

```diff
@@ -0,0 +1,192 @@
+      jfrStackTrace.put(THREAD_STATE, ev.getString(STATE));
+    }
+    jfrStackTrace.put(STACK_TRACE, MethodSupport.serialize(ev.getStackTrace()));
+    JvmStackTraceEvent event = stackTraceToStackFrames(jfrStackTrace);
+
+    stackTraceEventPerThread.computeIfPresent(
+        event.getThreadName(),
+        (k, list) -> {
+          list.add(event);
+          return list;
+        });
+
+    stackTraceEventPerThread.computeIfAbsent(
+        event.getThreadName(), k -> new ArrayList<>(Arrays.asList(event)));
+  }
+
+  @Override
+  public Stream<Event> summarize() {
+    Map<String, StackFrame> stackFrameByThread =
+        stackTraceEventPerThread
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey, event -> stackTraceToStackFrame(event.getValue())));
+
+    Map<String, List<FlameLevel>> flameLevelsByThread =
+        stackFrameByThread
+            .entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey, stackframe -> flattener.flatten(stackframe.getValue())));
+
+    List<Event> events =
+        flameLevelsByThread
+            .entrySet()
+            .stream()
+            .flatMap(e -> flameLevelToEvent(e.getValue(), e.getKey()).stream())
+            .collect(Collectors.toList());
+    return events.stream();
+  }
+
+  private List<Event> flameLevelToEvent(List<FlameLevel> flameLevels, String threadName) {
+    List<Event> events = new ArrayList<>();
+    for (FlameLevel flameLevel : flameLevels) {
+      Attributes attr = new Attributes();
+      attr.put(THREAD_NAME, threadName);
+      attr.put(FLAME_NAME, flameLevel.getName());
+      attr.put(FLAME_VALUE, flameLevel.getCount());
+      attr.put(FLAME_PARENT_ID, flameLevel.getParentId());
+      events.add(new Event(JFR_FLAMELEVEL, attr, timestamp.longValue()));
+    }
+    return events;
+  }
+
+  private StackFrame stackTraceToStackFrame(List<StackTraceEvent> traces) {
+    FlamegraphMarshaller out = new FlamegraphMarshaller();
+    for (StackTraceEvent event : traces) {
+      if (event instanceof JvmStackTraceEvent) {
+        JvmStackTraceEvent trace = (JvmStackTraceEvent) event;
+        Stack<String> stack = new Stack<>();
+        trace.getFrames().forEach(f -> stack.push(getFrameName(f)));
+        out.processEvent(stack, getValue());
+      }
+    }
+    return out.getStackFrame();
+  }
+
+  private Integer getValue() {
+    // For CPU stack traces this is always 1 - for other things (mem) it's not
+    return 1;
+  }
+
+  private String getFrameName(JvmStackTraceEvent.JvmStackFrame frame) {
+    StringBuilder entryBuilder = new StringBuilder();
+    entryBuilder.append(frame.getDesc());
+    entryBuilder.append(":");
+    entryBuilder.append(frame.getLine());
+    return entryBuilder.toString();
+  }
+
+  private static JvmStackTraceEvent stackTraceToStackFrames(Map<String, String> jfrStackTrace) {
```

Comment:
Should this, and similar methods, live on this class? Or would they be better elsewhere and with less restricted visibility?